src/loader.ts

```diff
@@ -1,207 +1,101 @@
-    version = getVersion(urlParts)(1);
-  }
-  return {
-    file,
-    name,
-    version,
-  };
-};
-
-/**
- * constructCachePath
- * @description a convenience function to construct the cache path
- * @param cache string
- * @param modulePath string
- * @returns string
- */
-export const constructCachePath = ({
-  cache,
-  modulePath,
-  debug = false,
-}: ConstructCachePath) => {
-  const { name, version, file } = getPackageNameVersionFromUrl(modulePath, debug);
-  if (debug) console.log(`${name}@${version}`);
-  return join(cache, `${name}@${version}`, file);
-}
+const config = processCliArgs(process.argv) || {};
+const initialCacheMap = createCacheMap(config?.values?.debug);
 
 /**
  * resolve
  * @description a convenience function to resolve the modules, this function is called by the loader automatically
- * @param specifier string
- * @param context object
- * @param nextResolve callback
- * @returns string
+ * @param {string} specifier
+ * @param {object} context
+ * @param {callback} nextResolve
+ * @returns {function} nextResolve
  */
-export const resolve = async (specifier: string, { parentURL }: Context, nextResolve: NextResolve) => {
-  const { nodeImportMapPath } = constructLoaderConfig()
-  if (!parentURL || !nodeImportMapPath) return nextResolve(specifier);
 
-    const importmap = constructImportMap(nodeImportMapPath)
-    const modulePath = importmap.resolve(
-      specifier,
-      cacheMap.get(parentURL) || parentURL
-    );
-    const moduleCachePath = await parseModule(specifier, modulePath);
-    return nextResolve(moduleCachePath);
-};
-
-/**
- * parseModule
- * @description a convenience function to parse the modules, this function is called by the loader automatically
- * @param specifier string
- * @param modulePath string
- * @returns string
- */
-export const parseModule = async (specifier: string, modulePath: string) => {
-  const { cache } = constructLoaderConfig()
-  const { protocol } = new URL(modulePath);
-  const isNode = protocol === "node:";
-  const isFile = protocol === "file:";
-
-  if (isNode || isFile) return specifier;
-
-  const cachePath = constructCachePath({ cache, modulePath })
-
-  cacheMap.set(`file://${cachePath}`, modulePath);
-  if (existsSync(cachePath)) return cachePath
-  const code = await (await fetch(modulePath).then(response => {
-    if (!response.ok) {
-      throw Error(`404: Module not found: ${modulePath}`);
-    }
-    return response;
-  })).text();
-
-  ensureFileSync(cachePath);
-  writeFileSync(cachePath, code);
-
-  return cachePath;
+export const resolve = async (
+  specifier: string,
+  { parentURL }: Context,
+  nextResolve: NextResolve,
+  options: ResolveOptions = config?.values,
+) => {
+  const { basePath, cachePath, debug: isDebugging = false, importmapPath, cacheMap = initialCacheMap } = options || {};
+  log.setLogger(isDebugging);
+  try {
+    // define importmap path
+    const cwd = process.cwd();
```

Comment:
This can be taken out, Anything inside the `resolve` is ran for every time a `import` statement is scanned in the module. So, we are just reading the same values multiple times.